File: Easy/nth smallest/BinarySearchTree.java
// Question:
// Write a Java program to identify the kth smallest element in the binary search tree where k =8.

import java.util.*;
import java.io.*;

class Node {
	int data;
	Node left, right;
	Node(int x)
	{
		data = x;
		left = right = null;
	}
}

public class BinaryTree {
	
	static int count = 0;
	public static Node insert(Node root, int x)
	{
		if (root == null)
			return new Node(x);
		if (x < root.data)
			root.left = insert(root.left, x);
		else if (x > root.data)
			root.right = insert(root.right, x);
		return root;
	}
	
	public static Node kthSmallest(Node root, int k)
	{
		if (root == null)
			return null;
		Node left = kthSmallest(root.left, k);
	
		if (left != null)
			return left;
			count++;
		if (count == k)
			return root;
	
		return kthSmallest(root.right, k);
	}
	
	public static void printKthSmallest(Node root, int k)
	{
		count = 0;
		
		Node res = kthSmallest(root, k);
		if (res == null)
			System.out.println("can't print as k is greater than nodes");
		else
			System.out.println(k+"-th Smallest Element is " + res.data);
	}
		
	public static void main (String[] args) {
		Scanner take = new Scanner(System.in);
		Node root = null;
        System.out.print("enter number of elements in the list: ");
		int size = take.nextInt();
	
        while(size-- !=0)
			root = insert(root, take.nextInt());
		System.out.print("enter k: ");
		int k = take.nextInt();
		printKthSmallest(root, k);
		
		
	}
}

File: ALGORITHMS/Tree/Binary_Search_Tree/Binary_search_tree.py

#a program used to demonstrate simple operations in a binary search tree

class Node:
    def __init__(self,node):
        self.left=None
        self.right=None
        self.node=node
    def insert(self,objnode):
        if self.node:
            if objnode<self.node:
                if self.left is None:
                    self.left=Node(objnode)
                else:
                    self.left.insert(objnode)
            elif objnode>self.node:
                if self.right is None:
                    self.right=Node(objnode)
                else:
                    self.right.insert(objnode)
            else:
                self.node=objnode
    def printBST(self):
        if self.left:
            self.left.printBST()
        print(self.node)
        if self.right:
            self.right.printBST()
    def Search(self,root, key):
        while root != None:
            
            # pass right subtree as new tree
            if key > root.node:
                root = root.right
    
            # pass left subtree as new tree
            elif key < root.node:
                root = root.left
            else:
                return True # if the key is found return 1
        return False        
            
n=int(input('Enter your Root node'))
root=Node(n)
while 1:
    a=int(input("enter the node you want to insert"))
    root.insert(a)
    ch=input("Do you want to continue Y or N")
    if ch=="Y"or ch=="y":
        continue
    else:
        break
print("Your final binary search tree is\n")
root.printBST()
x=int(input("Enter key to search for"))
if root.Search(root, x):
        print("Yes key exists in the tree")
else:
    print("Not found!")

